src/cdk-experimental/drag-drop/drag.ts

```diff
@@ -0,0 +1,493 @@
+      return new Promise(resolve => {
+        const handler = (event: Event) => {
+          if (event.target === this._preview) {
+            this._preview.removeEventListener('transitionend', handler);
+            resolve();
+          }
+        };
+
+        this._preview.addEventListener('transitionend', handler);
+      });
+    });
+  }
+
+  /**
+   * Sets the `transform` style on an element.
+   * @param element Element on which to set the transform.
+   * @param x Desired position of the element along the X axis.
+   * @param y Desired position of the element along the Y axis.
+   */
+  private _setTransform(element: HTMLElement, x: number, y: number) {
+    element.style.transform = `translate3d(${x}px, ${y}px, 0)`;
+  }
+
+  /**
+   * Helper to remove an element from the DOM and to do all the necessary null checks.
+   * @param element Element to be removed.
+   */
+  private _removeElement(element: HTMLElement | null) {
+    if (element && element.parentNode) {
+      element.parentNode.removeChild(element);
+    }
+  }
+
+  /** Removes the global event listeners that were bound by this draggable. */
+  private _removeDocumentEvents() {
+    this._document.removeEventListener('mousemove', this._pointerMove,
+      activeEventOptions as any);
+    this._document.removeEventListener('touchmove', this._pointerMove,
+      activeEventOptions as any);
+    this._document.removeEventListener('mouseup', this._pointerUp);
+    this._document.removeEventListener('touchend', this._pointerUp);
+  }
+
+  /** Determines the point of the page that was touched by the user. */
+  private _getPointerPositionOnPage(event: MouseEvent | TouchEvent): Point {
+    return this._isTouchEvent(event) ? {x: event.touches[0].pageX, y: event.touches[0].pageY} :
+                                       {x: event.pageX, y: event.pageY};
+  }
+
+  /** Listener used to prevent `touchmove` events while the element is being dragged. */
+  private _preventScrollListener = (event: TouchEvent) => {
+    if (this._isDragging) {
+      event.preventDefault();
+    }
+  }
+
+  /** Determines whether an event is a touch event. */
+  private _isTouchEvent(event: MouseEvent | TouchEvent): event is TouchEvent {
+    return event.type.startsWith('touch');
+  }
+
+  /** Destroys the preview element and its ViewRef. */
+  private _destroyPreview() {
+    if (this._preview) {
+      this._removeElement(this._preview);
+    }
+
+    if (this._previewRef) {
+      this._previewRef.destroy();
+    }
+
+    this._preview = this._previewRef = null!;
+  }
+
+  /** Destroys the placeholder element and its ViewRef. */
+  private _destroyPlaceholder() {
+    if (this._placeholder) {
+      this._removeElement(this._placeholder);
+    }
+
+    if (this._placeholderRef) {
+      this._placeholderRef.destroy();
+    }
+
+    this._placeholder = this._placeholderRef = null!;
+  }
+}
+
+/** Point on the page or within an element. */
+interface Point {
```

Comment:
This same interface is created and used in `flexible-connected-position-strategy.ts`, perhaps it should be moved to a common place?

Reply:
I decided to duplicate it since it's fairly small.